src/ucp/rma/flush.c

```diff
@@ -301,44 +301,48 @@ ucs_status_ptr_t ucp_ep_flush_internal(ucp_ep_h ep, unsigned uct_flags,
         return NULL;
     }
 
-    req = ucp_request_get(ep->worker);
-    if (req == NULL) {
-        return UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-    }
+    req = ucp_request_get_param(ep->worker, param,
+                                {return UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);});
 
     /*
-     *  Flush operation can be queued on the pending queue of only one of the
+     * Flush operation can be queued on the pending queue of only one of the
      * lanes (indicated by req->send.lane) and scheduled for completion on any
      * number of lanes. req->send.uct_comp.count keeps track of how many lanes
      * are not flushed yet, and when it reaches zero, it means all lanes are
      * flushed. req->send.flush.lanes keeps track of which lanes we still have
      * to start flush on.
-      */
-    req->flags                    = req_flags;
-    req->status                   = UCS_OK;
-    req->send.ep                  = ep;
-    req->send.cb                  = (ucp_send_nbx_callback_t)req_cb;
-    req->send.flush.flushed_cb    = flushed_cb;
-    req->send.flush.prog_id       = UCS_CALLBACKQ_ID_NULL;
-    req->send.flush.uct_flags     = uct_flags;
-    req->send.flush.worker_req    = worker_req;
-    req->send.flush.sw_started    = 0;
-    req->send.flush.sw_done       = 0;
-    req->send.flush.num_lanes     = ucp_ep_num_lanes(ep);;
-    req->send.flush.started_lanes = 0;
-
-    req->send.lane              = UCP_NULL_LANE;
-    req->send.uct.func          = ucp_ep_flush_progress_pending;
-    req->send.state.uct_comp.func   = ucp_ep_flush_completion;
-    req->send.state.uct_comp.count  = ucp_ep_num_lanes(ep);
+     */
+    req->flags                     = req_flags;
+    req->status                    = UCS_OK;
+    req->send.ep                   = ep;
+    req->send.flush.flushed_cb     = flushed_cb;
+    req->send.flush.prog_id        = UCS_CALLBACKQ_ID_NULL;
+    req->send.flush.uct_flags      = uct_flags;
+    req->send.flush.worker_req     = worker_req;
+    req->send.flush.sw_started     = 0;
+    req->send.flush.sw_done        = 0;
+    req->send.flush.num_lanes      = ucp_ep_num_lanes(ep);;
+    req->send.flush.started_lanes  = 0;
+
+    req->send.lane                 = UCP_NULL_LANE;
+    req->send.uct.func             = ucp_ep_flush_progress_pending;
+    req->send.state.uct_comp.func  = ucp_ep_flush_completion;
+    req->send.state.uct_comp.count = ucp_ep_num_lanes(ep);
+
+    if (param->op_attr_mask & UCP_OP_ATTR_FIELD_CALLBACK) {
+        ucp_request_set_callback(req, send.cb, param->cb.send,
```

Comment:
can use `ucp_request_set_send_callback_param` here

Reply:
yep, why not :)